python/tvm/contrib/target/android_nnapi/relayir_to_nnapi_converter/export_object.py

```diff
@@ -0,0 +1,304 @@
+            self.helper.type_to_idx_map[tipe] = len(self["types"]) - 1
+        return self.helper.type_to_idx_map[tipe]
+
+    @staticmethod
+    def _canonicalize_scalar_constant(dtype, val):
+        # skip canonicalizing strings as they may carry specific meanings,
+        # e.g. macro-defined values
+        if not isinstance(val, str):
+            if dtype == "float16":
+                if isinstance(val, float):
+                    val = hex(
+                        struct.unpack("H", struct.pack("e", val))[0]
+                    )  # for float16 we use uint16_t in C, hence the conversion
+            elif dtype == "float32":
+                val = float(val)
+            elif dtype == "int32":
+                val = int(val)
+            elif dtype == "uint32":
+                val = int(val)
+            elif dtype == "bool":
+                val = bool(val)
+            else:
+                assert False, "Unreachable"
+        return val
+
+    def add_scalar_constant(self, val, dtype):
+        """Add scalar constant to export object
+
+        Parameters
+        ----------
+        val: numerical or str
+            value of the constant. Can be defined constant in the NNAPI framework.
+
+        dtype: str
+            data type of the constant
+
+        Returns
+        -------
+        index: int
+            index of the constant in export object constants array
+        """
+        # canonicalize
+        dtype = str(dtype)
+        assert_anc_compatibility(
+            dtype in ["float16", "float32", "int32", "uint32", "bool"],
+            f"Unsupported data type { dtype }",
+        )
+        val = self._canonicalize_scalar_constant(dtype, val)
+
+        new_const = {
+            "type": "scalar",
+            "dtype": dtype,
+            "value": val,
+        }
+        if new_const in self["constants"]:
+            return self["constants"].index(new_const)
+
+        self["constants"].append(new_const)
+        return len(self["constants"]) - 1
+
+    def add_array_constant(self, vals, dtype):
+        """Add array constant to export object
+
+        Parameters
+        ----------
+        vals: array of values in dtype
+            values of array
+
+        dtype: string
+            data type of array
+
+        Returns
+        -------
+        index: int
+            index of added constant in export_obj["constants"]
+        """
+        # canonicalize
+        dtype = str(dtype)
+        assert_anc_compatibility(
+            dtype in ["float16", "float32", "int32", "uint32", "bool"],
+            f"Unsupported data type { dtype }",
+        )
+        assert len(vals) > 0, "Array constant should not be empty"
+        vals = list(map(lambda v: self._canonicalize_scalar_constant(dtype, v), vals))
+
+        new_const = {
+            "type": "array",
+            "dtype": dtype,
+            "value": vals,
+        }
+        if new_const in self["constants"]:
```

Comment:
Is this correct? `new_const (dict)` should not be hashable?

Reply:
`self["constants"]` is a list, so I think it's fine? (`new_const` is not used as key)

Reply:
I see. It should be fine then.